### PR TITLE
nixos/pgbouncer: add services.pgbouncer.settings option

### PR DESCRIPTION
**Pull request by tie**
## Description of changes

This change adds services.pgbouncer.settings option as per [RFC 0042] and removes other options that were previously used to generate configuration file.

In addition to that, we also place the configuration file under environment.etc to allow reloading configuration without service restart.

[RFC 0042]: https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by tie** [comment]
W.r.t. `notify-reload` vs `notify`:

Systemd docs: https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#:~:text=Behavior%20of%20notify-reload,reloading%20of%20the%20service's%20configuration.

Relevant pgbouncer code: https://github.com/pgbouncer/pgbouncer/blob/6c629ab98e8c06aec05feda49019908693121755/src/main.c#L546-L554

**Comment by tie** [comment]
@wegank, sorry for pinging you directly, but since we already have some potentially breaking changes in `pgbouncer` module configuration for 24.05 release, is it possible to get this PR reviewed and hopefully merged before release? Or it’s already too late in 24.05 release schedule?

**Comment by wegank** [comment]
Sounds too late for me. Also @Mic92.

**Comment by tie** [comment]
OK, that’s fine, thanks, just wanted to confirm.

File: nixos/doc/manual/release-notes/rl-2405.section.md [hunk 1]
```diff
@@ -412,7 +412,7 @@ The pre-existing [services.ankisyncd](#opt-services.ankisyncd.enable) has been m
   `wants`), because the dependency that `multi-user.target` has on
   `network-online.target` is planned for removal.
 
-- `services.pgbouncer` now has systemd support enabled and will log to journald. The default setting for `services.pgbouncer.logFile` is now `null` to disable logging to a separate log file.
+- `services.pgbouncer` now has systemd support enabled and will log to journald. The systemd service is configured with `Type=notify-reload` and allows reloading configuration without process restart. PgBouncer configuration options were moved to the free-form type option named [`services.pgbouncer.settings`](#opt-services.pgbouncer.settings) according to the NixOS RFC 0042.
```
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2405.section.md:415]:
```suggestion
- `services.pgbouncer` now has systemd support enabled and will log to journald.
  The systemd service is configured with `Type=notify-reload` and allows reloading configuration without process restart.
  PgBouncer configuration options were moved to the free-form type option named [`services.pgbouncer.settings`](#opt-services.pgbouncer.settings) according to the NixOS RFC 0042.
```

File: nixos/modules/services/monitoring/prometheus/exporters/pgbouncer.nix [hunk 2]
```diff
@@ -116,35 +120,24 @@ in
...
-            ++ optionals (cfg.webConfigFile != null) [
-            "--web.config.file ${escapeShellArg cfg.webConfigFile}"
-          ]
-            ++ cfg.extraFlags)}";
-      in
-      {
-        ExecStart = "${startScript}/bin/pgbouncer-start";
-      };
+    script = concatStringsSep " " ([
+      "${pkgs.prometheus-pgbouncer-exporter}/bin/pgbouncer_exporter"
```
Comment (SuperSandro2000) [inline nixos/modules/services/monitoring/prometheus/exporters/pgbouncer.nix:124]:
```suggestion
      "${lib.getExe pkgs.prometheus-pgbouncer-exporter}"
```

File: nixos/doc/manual/release-notes/rl-2405.section.md (hunk 1 above)
Comment (tie) [reply nixos/doc/manual/release-notes/rl-2405.section.md:415]: 👍
Also we should probably wait for 24.11 release notes file to be initialized, see above w.r.t. release schedule.

File: nixos/modules/services/monitoring/prometheus/exporters/pgbouncer.nix (hunk 2 above)
Comment (tie) [reply nixos/modules/services/monitoring/prometheus/exporters/pgbouncer.nix]: Done. Also added package option.

File: nixos/modules/services/databases/pgbouncer.nix [hunk 3]
```diff
@@ -1,543 +1,58 @@
...
-    };
-
-    # Authentication settings
-    authType = mkOption {
-      type = types.enum [ "cert" "md5" "scram-sha-256" "plain" "trust" "any" "hba" "pam" ];
-      default = "md5";
+    settings = lib.mkOption {
+      type = settingsFormat.type;
+      default = { };
       description = ''
```
[the remark this answers is not in the document]
Comment (tie) [reply nixos/modules/services/databases/pgbouncer.nix:53]: How do we decide which settings deserve to show up in search? In addition to that, I assume that people that need PgBouncer are actually familiar with all the options, and non-NixOS-specific options add a lot of noise to the search results (at least that is the case for me in current pgbouncer module implementation).

File: nixos/modules/services/databases/pgbouncer.nix [hunk 4]
```diff
@@ -1,543 +1,58 @@
...
-  };
-
-in {
+  settingsFormat = pkgs.formats.ini { };
+  configFile = settingsFormat.generate "pgbouncer.ini" cfg.settings;
+  configPath = "pgbouncer/pgbouncer.ini";
+in
+{
+  imports = [
+    (lib.mkRemovedOptionModule [ "services" "pgbouncer" "logFile" ] "Use services.pgbouncer.settings instead.")
```
[the remark this answers is not in the document]
Comment (tie) [reply nixos/modules/services/databases/pgbouncer.nix]: Done.

File: nixos/doc/manual/release-notes/rl-2405.section.md (hunk 1 above)
Comment (tie) [reply nixos/doc/manual/release-notes/rl-2405.section.md]: Moved to 24.11 release notes.

**Review by Ma27 — COMMENTED** [review]
I must say, I don't like this change very much.

RFC42 encourages to use freeform modules, i.e. predefined options for type-checking and defaults in e.g. the manual.

Even worse, two pitfalls when configuring pgbouncer that were previously caught with assertions are now gone and people will find out at runtime.

All of taht can be added in an RFC42-compliant module, but for me, the current state makes the module worse than it was before.

**Comment by tie** [comment]
>Even worse, two pitfalls when configuring pgbouncer that were previously caught with assertions are now gone and people will find out at runtime.

If you are talking about pgbouncer-exporter assertions in prometheus module, I think these are a bad design. The evaluated configuration is not necessarily the one being monitored. It’s OK to delay some issues until run time if we can’t check for them at evaluation time (or it’s not possible to do that efficiently).

>RFC42 encourages to use freeform modules, i.e. predefined options for type-checking and defaults in e.g. the manual.

That’s not how I understand [RFC42 Part 2](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md#part-2-1), but I may be wrong here. I agree that we should add common predefined settings for the pgbouncer module. I didn’t do that initially (https://github.com/NixOS/nixpkgs/pull/308700#discussion_r1600285914), but I’ll add these in a subsequent PR soon, in particular, `listen_addr`, `listen_port` and `unix_socket_dir` (since we are using these in the configuration and these are the most basic options for pgbouncer setup).

File: nixos/modules/services/databases/pgbouncer.nix [hunk 5]
```diff
@@ -1,543 +1,138 @@
...
+  imports = [
+    (lib.mkRenamedOptionModule
+      [ "services" "pgbouncer" "logFile" ]
+      [ "services" "pgbouncer" "settings" "pgbouncer" "log_file" ])
+    (lib.mkRenamedOptionModule
+      [ "services" "pgbouncer" "listenAddress" ]
+      [ "services" "pgbouncer" "settings" "pgbouncer" "listen_addr" ])
+    (lib.mkRenamedOptionModule
+      [ "services" "pgbouncer" "listenPort" ]
+      [ "services" "pgbouncer" "settings" "pgbouncer" "listen_port" ])
```
Comment (Ma27) [inline nixos/modules/services/databases/pgbouncer.nix:19]: This gives me

```
error: evaluation aborted with the following error message: 'Renaming error: option `services.pgbouncer.settings.pgbouncer.listen_port' does not exist.'
```

btw.
Keeping options around is also useful for a painless upgrade path.

**Comment by Ma27** [comment]
Opened #356965.

**Comment by omares** [comment]
Hey, it seems that pgbouncer handles reloads differently. This causes the daemon to hang indefinitely. You can see more details in issue #507 on GitHub: https://github.com/pgbouncer/pgbouncer/issues/507.

Re-adding the following code to my `systemd.services.pgbouncer` definition resolves the issue:

``` 
serviceConfig = {
    Type = "notify-reload";
    ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
}; 
```

Maybe the removal of the change should be reverted?

**Pull request merged**